controllers/drinks.js

```diff
@@ -0,0 +1,53 @@
+const Cocktail = require('../models/cocktail');
+const Category = require('../models/category');
+
+const getMainPage = async (req, res, next) => {
+  try {
+    const numberOfCocktails = Number(req.params.quantity);
+
+    if (
+      typeof numberOfCocktails !== 'number' ||
+      (numberOfCocktails !== 1 &&
+        numberOfCocktails !== 2 &&
+        numberOfCocktails !== 3)
+    ) {
+      res.status(400).json({ message: 'Quantity must be a number(1-3)!' });
+    }
+
+    const categories = await Category.find();
+
+    if (categories == null) {
+      res.status(404).json({ message: 'Categories not found' });
+    }
+
+    const randomCocktails = [];
+
+    function getRandomItemFromArray(arr, count) {
+      const shuffled = arr.sort(() => 0.5 - Math.random());
+      return shuffled.slice(0, count);
+    }
+
+    const randomCategories = getRandomItemFromArray(categories, 3);
+
+    for (const item of randomCategories) {
+      const cocktailsInCategory = await Cocktail.find({
+        category: item.category,
+      }).limit(numberOfCocktails);
+
+      if (cocktailsInCategory == null) {
+        res.status(404).json({ message: 'Cocktail not found' });
+      }
+
+      randomCocktails.push({
+        category: item.category,
+        cocktails: cocktailsInCategory,
+      });
+    }
+
+    res.send(randomCocktails);
+  } catch (error) {
+    console.log(error);
+  }
+};
+
+module.exports = { getMainPage };
```

Comment:
let discuss the following implementation:

const Drink = require('../models/drink');
const Category = require('../models/category');
const User = require('../models/user');
const getUserAge = require('../helpers/getUserAge');

const getRandomDrinks = async (req, res, next) => {
  let { drinksPerCategory } = req.query;

  drinksPerCategory = parseInt(drinksPerCategory);
  drinksPerCategory =
    isNaN(drinksPerCategory) || drinksPerCategory < 1 ? 3 : drinksPerCategory;

  try {
    const categories = await Category.find();

    if (!categories || categories.length === 0) {
      return res.status(404).json({ message: 'No categories found' });
    }

    const user = await User.findById(req.user.id);
    const userAge = getUserAge(user.dateOfBirth);

    let matchCondition = {};
    if (userAge < 18) {
      matchCondition = { alcoholic: 'Non alcoholic' };
    }

    const categoriesWithDrinks = await Promise.all(
      categories.map(async category => {
        const drinks = await Drink.aggregate([
          { $match: { category: category.category, ...matchCondition } },
          { $sample: { size: drinksPerCategory } },
        ]);
        return { category: category.category, drinks };
      })
    );

    return res.status(200).json(categoriesWithDrinks);
  } catch (error) {
    next(error);
  }
};

module.exports = { getRandomDrinks };

Reply:
create helpers/getUserAge.js:

function getUserAge(dateOfBirth) {
  const today = new Date();
  const birthDate = new Date(dateOfBirth);

  let age = today.getFullYear() - birthDate.getFullYear();
  const monthDifference = today.getMonth() - birthDate.getMonth();

  if (
    monthDifference < 0 ||
    (monthDifference === 0 && today.getDate() < birthDate.getDate())
  ) {
    age--;
  }

  return age;
}

module.exports = getUserAge;